client/src/main/scripts/components/search/database/SearchDatabaseApi.ts

```diff
@@ -20,148 +20,83 @@
-      if (!searchTerm) {
-        throw "SearchDatabaseApi: Must provide a searchTerm when creating an instance";
-      }
-      SearchDatabaseApi.instance = new SearchDatabaseApi(searchTerm);
-      return SearchDatabaseApi.instance;
-    }
-    if (
-      SearchDatabaseApi.instance.searchTerm &&
-      SearchDatabaseApi.instance.searchTerm !== searchTerm
-    ) {
-      throw "SearchDatabaseApi: cannot change search term";
-    }
-    return SearchDatabaseApi.instance;
+import { postRequestReturningPromise } from "../../../utils/databaseAPI";
+import { FoundationData } from "../../../utils/foundationData/FoundationData";
+import { FoundationDataBuilder } from "../../../utils/foundationData/FoundationDataBuilder";
+
+class SearchResult {
+  constructor(
+    private _factTurns: VideoFactsToTurns[],
+    private _searchTerm: string
+  ) {}
+
+  get factTurns() {
+    return this._factTurns;
   }
 
-  connect = (): Promise<void> => {
-    return new Promise((resolve) => {
-      const bodyJson: Search.Request = {
-        q: this.searchTerm,
-      };
-      postRequest(
-        Routes.API_SEARCH,
-        bodyJson,
-        (json: Search.FactResultList) => {
-          try {
-            this.getFacts(json, resolve);
-          } catch (err) {
-            reject(err);
-          }
-        }
-      );
-    });
-  };
-
-  get searchDatabase() {
-    if (!this._searchDatabase) {
-      throw "SearchDatabaseApi: Database must be loaded before it can be accessed. Ensure connect has been called.";
-    }
-    return this._searchDatabase;
+  get searchTerm() {
+    return this._searchTerm;
   }
+}
 
-  private getFacts(
-    factList: Search.FactResultList,
-    onSuccess: (value?: void | PromiseLike<void> | undefined) => void
-  ) {
-    const hashesToTurns = createHashMap(factList);
-    if (!hashesToTurns) {
-      this._searchDatabase = new SearchDatabase([], this.searchTerm);
-      return;
-    }
-    const innerPromises: Promise<VideoFactHashMap>[] = [];
-    hashesToTurns.forEach((turns, hash) => {
-      innerPromises.push(fetchFactReturningPromise(hash));
-    });
+// What's the advantage of this class instead of having the `search` function call `searchImpl` directly?
+class SearchWithData {
+  constructor(
+    public searchTerm: string,
+    public videoResults: Search.VideoResult[],
+    public foundationData: FoundationData
+  ) {}
+}
 
-    Promise.all(innerPromises).then((videoFacts: VideoFactHashMap[]) => {
-      const factTurns = processFacts(hashesToTurns, videoFacts);
-      this._searchDatabase = new SearchDatabase(factTurns, this.searchTerm);
-      onSuccess();
-    });
-  }
+export async function search(searchTerm: string): Promise<SearchResult> {
+  const bodyJson: Search.Request = {
+    q: searchTerm,
+  };
+  const factResults: Search.FactResultList = await postRequestReturningPromise(
+    Routes.API_SEARCH,
+    bodyJson
+  );
+
+  const builder = new FoundationDataBuilder(); // This should probably live elsewhere
```

Comment:
I think this is exactly the right place!

Reply:
But what if nGramLoader wants to do `builder.add` stuff? It will need to create a _new_ FoundationDataBuilder.